### PR TITLE
`Learning analytics`: Improve usability of student course dashboard

### PR DESCRIPTION
**Pull request by FelixTJDietrich**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] ~I added multiple integration tests (Spring) related to the features (with a high test coverage).~
- [x] ~I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).~
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] ~I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).~
- [x] ~I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).~
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently the student dashboard has some usability issues making it not as useful for students.
image (3)

### Description

This PR fixes the following regarding the student course dashboard:

- [x] Hide competency from dashboard when there are no released exercises or lecture units 
- [x] Move judgement of learning prompt to the top (call to action first)
- [x] Add precondition to judgement of learning to prompt it only if the student has at least made some progress (either in exercises or in lecture units)
- [x] Prefix lecture units with `<lecture title> - ` to provide more context for the lecture unit
- [x] Update the attachment unit icon to represent a file rather than a download icon
- [x] Added info tooltip to judgement of learning prompt for explanation 
- [x] Improve tooltips generally on the course dashboard
- [x] Renamed "Next objectives" to "Next activities to master this competency" for clarity
- [x] Show at most 5 next activities
    - [x] Move exercises before the lecture units since they are more important
    - [ ] ~Also show exercises of the past, if the student does not have 80% of `max(gradedScore, practiceScore)`~ -> Follow-up
    - [ ] ~Sort exercises with lowest score first so the student knows immediately where they can make the biggest learning impact~ -> Follow-up
    - [ ] ~Show “Congratulations, you finished all released units in this competency” if there is currently nothing to do for this competency~ -> Follow-up

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1-2 Students
- 1 Course with `Student Analytics Course Dashboard` enabled
- Some competencies associated with lecture units and exercises more than 5 
- Use TS1 with `Test Course Felix T.J. Dietrich`

1. Log in to Artemis
2. Navigate to `Test Course Felix T.J. Dietrich` on TS1 (or create your own course with dashboard enabled in the course settings)
3. Visit the dashboard and verify the described changes.
    1. It shows only max 5 items within a competency 
    2. Exercises are before lectures
    3. Lecture units are prefixed with the lecture title
4. Check some lectures in `Lectures` as complete
5. Go back to dashboard there should hopefully be a ( ? ) with a prompt in the competency to rate your mastery
   1. The "How would you rate your mastery" prompt should now be first (before lectures and exercises)
   2. There is an info (?) thingy explaining it
6. Verify that all info description (?) make sense, i.e. they are helpful and understandable
 

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Screenshots

image
image
image
image

## Summary by CodeRabbit

- **New Features**
  - Introduced a new rating component for judging learning in competencies.
  - Added lecture title visibility in lecture unit reports.

- **Improvements**
  - Enhanced competency progress checks for more accurate evaluations.
  - Improved display of next exercises and lecture units in the competency accordion.

- **Bug Fixes**
  - Adjusted mappings for competency metrics to ensure correct data relations.

- **Documentation**
  - Minor update to the Jest configuration coverage thresholds.

### CODE REVIEW

File: src/main/webapp/i18n/de/courseStudentDashboard.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
             },
             "exercisePerformance": {
                 "title": "Übungsleistungen",
-                "info": "Leistung wird als Prozentsatz der Punkte gemessen, die du in einer Aufgabe erreicht hast.",
+                "info": "Dies zeigt, wie gut du in jeder Aufgabe im Vergleich zu den maximal möglichen Punkten abgeschnitten hast. Du kannst auch das Durchschnittsergebnis aller Studierenden sehen, um deine Leistung mit deinen Kommilitonen zu vergleichen.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/courseStudentDashboard.json:10]:
```suggestion
                "info": "Dies zeigt, wie gut du in jeder Aufgabe im Vergleich zu den maximal möglichen Punkten abgeschnitten hast. Du kannst auch das Durchschnittsergebnis aller Studierenden sehen, um deine Leistung mit deinen Kommiliton:innen zu vergleichen.",
```

File: src/main/webapp/i18n/de/courseStudentDashboard.json [hunk 2]
```diff
@@ -16,15 +16,16 @@
             },
             "exerciseLateness": {
                 "title": "Übungsverspätungen",
-                "info": "Relative Verspätung bezieht sich auf den Prozentsatz der Arbeitszeit, die vergangen ist.",
+                "info": "Dies zeigt, wie viel Zeit du genutzt hast, um jede Aufgabe zu bearbeiten und einzureichen, als Prozentsatz der insgesamt erlaubten Zeit. Vergleiche deine Einreichungszeiten mit dem Durchschnitt, um zu sehen, ob du früher oder später als die meisten Studierenden einreichst.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/courseStudentDashboard.json:19]:
```suggestion
                "info": "Dies zeigt, wie viel Zeit du genutzt hast, um jede Aufgabe zu bearbeiten und einzureichen, als Prozentsatz der insgesamt verfügbaren Zeit. Vergleiche deine Einreichungszeiten mit dem Durchschnitt, um zu sehen, ob du früher oder später als die meisten Studierenden einreichst.",
```

File: src/main/webapp/i18n/en/courseStudentDashboard.json [hunk 3]
```diff
@@ -16,7 +16,7 @@
             },
             "exerciseLateness": {
                 "title": "Exercise Lateness",
-                "info": "Relative lateness refers to the percentage of working time that has passed.",
+                "info": "This shows how much time you used to complete and submit each exercise as a percentage of the total time allowed. Compare your submission times with the average to see if you're earlier or later than most students.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/en/courseStudentDashboard.json:19]:
```suggestion
                "info": "This shows how much time you used to complete and submit each exercise as a percentage of the total time available. Compare your submission times with the average to see if you're earlier or later than most students.",
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/i18n/de/courseStudentDashboard.json (hunk 1 above)
Comment (FelixTJDietrich) [reply src/main/webapp/i18n/de/courseStudentDashboard.json:10]: Used `anderen Studierenden`

**Review by bassner — APPROVED** [review]
Tested on TS1. The listed changes work. Code LGTM 👍🏽

**Review by JohannesWt — APPROVED** [review]
The following is probably unrelated to this PR but important to mention:
- clicking on "Competency without Taxonomy" (same with Taxonomy)
    - clicking on "View All lectures and Exercises"
    - loading spinner on Competencies page is not disappearing
- also the prog exercises are not shown on the competencies page 

The rest works as described. 👍

**Review by marlon-luca-bu — COMMENTED** [review]
Tested on TS1, everything works as expected. So besides the minor things, looks good 👍

**Review by kaancayli — APPROVED** [review]
Tested on TS1. Could verify the changes. Works as expected.

**Comment by EneaGore** [comment]
Tested on TS1. English version looks good and changes too. Not sure if this screenshot is related to this PR
Screenshot 2024-06-18 154103

**Review by EneaGore — APPROVED** [review]

**Comment by FelixTJDietrich** [comment]
> Tested on TS1. English version looks good and changes too. Not sure if this screenshot is related to this PR Screenshot 2024-06-18 154103

Will be fixed in another PR

**Comment by FelixTJDietrich** [comment]
> The following is probably unrelated to this PR but important to mention:
> 
> * clicking on "Competency without Taxonomy" (same with Taxonomy)
>   
>   * clicking on "View All lectures and Exercises"
>   * loading spinner on Competencies page is not disappearing
> * also the prog exercises are not shown on the competencies page
> 
> The rest works as described. 👍 (Tested on TS1)

I can't reproduce this issue and I think it is probably unrelated to this PR

File: src/main/java/de/tum/in/www1/artemis/repository/metrics/LectureUnitMetricsRepository.java [hunk 4]
```diff
@@ -27,7 +27,7 @@ public interface LectureUnitMetricsRepository extends JpaRepository<LectureUnit,
      * @return the lecture unit information for all lecture units in the course
      */
     @Query("""
-            SELECT new de.tum.in.www1.artemis.web.rest.dto.metrics.LectureUnitInformationDTO(lu.id, lu.lecture.id, COALESCE(lu.name, a.name), lu.releaseDate, TYPE(lu))
+            SELECT new de.tum.in.www1.artemis.web.rest.dto.metrics.LectureUnitInformationDTO(lu.id, lu.lecture.id, lu.lecture.title, COALESCE(lu.name, a.name), lu.releaseDate, TYPE(lu))
```
[the remark this answers is not in the document]
Comment (FelixTJDietrich) [reply src/main/java/de/tum/in/www1/artemis/repository/metrics/LectureUnitMetricsRepository.java:30]: Will keep it as is.

File: src/main/webapp/i18n/de/courseStudentDashboard.json [hunk 5]
```diff
@@ -16,15 +16,16 @@
...
                 "noData": "Du hast noch keine Aufgaben in diesem Kurs eingereicht.",
                 "yourLatenessLabel": "Deine Verspätung",
                 "averageLatenessLabel": "Durchschnittliche Verspätung",
                 "xAxisLabel": "Aufgaben",
                 "yAxisLabel": "Relative Fristnähe (%)"
             },
             "judgementOfLearning": {
-                "title": "Wie würdest du dein Verständnis bewerten?",
+                "title": "Wie würdest du deine Kompetenzbeherrschung bewerten?",
+                "info": "Bewerte, wie gut du die Fertigkeiten aus dieser Kompetenz verstehst und anwenden kannst. Artemis bietet dir auch eine Bewertung basierend auf deiner Leistung, damit du siehst, wo du stehst.",
```
[the remark this answers is not in the document]
Comment (FelixTJDietrich) [reply src/main/webapp/i18n/de/courseStudentDashboard.json:28]: Will keep it as is.

File: src/main/webapp/i18n/de/courseStudentDashboard.json [hunk 6]
```diff
@@ -16,15 +16,16 @@
             },
             "exerciseLateness": {
                 "title": "Übungsverspätungen",
-                "info": "Relative Verspätung bezieht sich auf den Prozentsatz der Arbeitszeit, die vergangen ist.",
+                "info": "Dies zeigt, wie viel Zeit du genutzt hast, um jede Aufgabe zu bearbeiten und einzureichen, als Prozentsatz der insgesamt verfügbaren Zeit. Vergleiche deine Einreichungszeiten mit dem Durchschnitt, um zu sehen, ob du früher oder später als die meisten Studierenden einreichst.",
```
[the remark this answers is not in the document]
Comment (FelixTJDietrich) [reply src/main/webapp/i18n/de/courseStudentDashboard.json:19]: Will keep it as is.

**Review by sarpsahinalp — APPROVED** [review]
Tested on ts1 listed changes works

**Review by laadvo — COMMENTED** [review]
Tested on ts1, the student analytics dashboard is working as expected :+1:

**Review by laadvo — APPROVED** [review]
Approved

**Review by DominikRemo — APPROVED** [review]
Code LGTM

**Pull request merged**